crates/red_knot/src/check.rs

```diff
@@ -0,0 +1,45 @@
+use crate::files::{FileId, Files};
+use crate::Workspace;
+use parking_lot::lock_api::RwLockUpgradableReadGuard;
+use parking_lot::RwLock;
+use rustc_hash::FxHashMap;
+use std::collections::HashMap;
+
+/// The basic idea of Salsa is that you give it all inputs (the source contents) and
+/// you can then define queries on top of it.
+///
+/// That means we need to eagerly set the contents of all files that the analysis could reach, which means
+/// we possibly need to load the entire program into memory, which is wasteful.
+/// TODO: I don't want to eagerly load all the files. We should load them lazily.
+/// TODO: Salsa also has no concept of a persistent cache
```

Comment:
This could be an issue. In looking at the six year old salsa issue for persistent cache, I found this comment from someone with experience building these kinds of tools that recommends against persistent caches entirely: https://github.com/rust-lang/rfcs/pull/1317#issuecomment-150965895

I’m not convinced, though, particularly for the CLI use case where laziness doesn’t help much.

Reply:
Yeah, I think this is different for a CLI where you e.g. want to check your entire program. I think rust-analyzer does some indexing during startup and that can easily take 10 seconds. I don't think our users would appreciate it if Ruff suddenly takes 10s in CI.